Document context refacto and modifications regarding symfony controllers

jolelievre opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop developer documentation! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines on how to contribute:
https://devdocs.prestashop-project.org/8/contribute/documentation/how/
------------------------------------------------------------------------------>

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 9.x
| Description?  | Document context refacto and modifications regarding symfony controllers
| Fixed ticket? | ~

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

jolelievre committed:
Document context refacto and modifications regarding symfony controllers

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -78,7 +78,91 @@ Some dependencies are still present but were upgraded which comes with their own
 ### Symfony controllers
 
-@todo: explain changes about DI (controllers as services, injected container no longer the same, ...), new PrestaShopAdminController and deprecated FrameworkAdminBundleController
+In Symfony 6.4 the controllers must now be defined as services, the impact for the existing controllers is mainly around the concept od Dependency Injection and how services are injected or accessed in the controllers.
```
```suggestion
In Symfony 6.4, the controllers must now be defined as services. The impact on the existing controllers is mainly around the concept of Dependency Injection and how services are injected or accessed in the controllers.
```

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -78,7 +78,91 @@ Some dependencies are still present but were upgraded which comes with their own
 
-@todo: explain changes about DI (controllers as services, injected container no longer the same, ...), new PrestaShopAdminController and deprecated FrameworkAdminBundleController
+In Symfony 6.4 the controllers must now be defined as services, the impact for the existing controllers is mainly around the concept od Dependency Injection and how services are injected or accessed in the controllers.
+The container that is passed in the controllers is no longer the "global container" that contains all the existing services in the application, instead it injects a dedicated container optimized for the controller based on the services that are injected into it.
```
```suggestion
The container passed to the controllers is no longer the "global container" that contains all the existing services in the application. Instead, it injects a dedicated container optimized for the controller based on the services that are injected into it.
```

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -78,7 +78,91 @@ Some dependencies are still present but were upgraded which comes with their own
+In Symfony 6.4 the controllers must now be defined as services, the impact for the existing controllers is mainly around the concept od Dependency Injection and how services are injected or accessed in the controllers.
+The container that is passed in the controllers is no longer the "global container" that contains all the existing services in the application, instead it injects a dedicated container optimized for the controller based on the services that are injected into it.
+
+The problem is that most of our Symfony controllers don't rely on injection but use the `$this->get('service_name')` method to get any service, this is no longer possible in modern controllers because of the mentioned optimisation related to container build.
```
```suggestion
The issue is that most of PrestaShop's Symfony controllers do not rely on injection but instead use the `$this->get('service_name')` method to access any service. This is no longer possible in modern controllers because of the mentioned optimization related to container build.
```

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -78,7 +78,91 @@ Some dependencies are still present but were upgraded which comes with their own
+The container that is passed in the controllers is no longer the "global container" that contains all the existing services in the application, instead it injects a dedicated container optimized for the controller based on the services that are injected into it.
+
+The problem is that most of our Symfony controllers don't rely on injection but use the `$this->get('service_name')` method to get any service, this is no longer possible in modern controllers because of the mentioned optimisation related to container build.
+To avoid a very big breaking change we modified the `PrestaShopBundle\Controller\Admin\FrameworkBundleAdminController` so that it is able to fetch services from the controller's container AND from the global one, but this is a little hack that goes against Symfony recommendations, as such this base class is already **deprecated** and will be removed in PrestaShop 10.0.
```
```suggestion
To avoid a very big breaking change, we modified the `PrestaShopBundle\Controller\Admin\FrameworkBundleAdminController`, so that it can fetch services from the controller's container **and** from the global one, but this is a little hack that goes against Symfony recommendations, as such this base class is already **deprecated and will be removed in PrestaShop 10.0**.
```

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -78,7 +78,91 @@ Some dependencies are still present but were upgraded which comes with their own
+The problem is that most of our Symfony controllers don't rely on injection but use the `$this->get('service_name')` method to get any service, this is no longer possible in modern controllers because of the mentioned optimisation related to container build.
+To avoid a very big breaking change we modified the `PrestaShopBundle\Controller\Admin\FrameworkBundleAdminController` so that it is able to fetch services from the controller's container AND from the global one, but this is a little hack that goes against Symfony recommendations, as such this base class is already **deprecated** and will be removed in PrestaShop 10.0.
+
+This gives developer some time to refacto their controller, they should now rely on the new `PrestaShopBundle\Controller\Admin\PrestaShopAdminController` which doesn't implement this hack in order to force a proper implementation of the controllers.
```
```suggestion
This should give developers some time to refacto their controllers. They should now rely on the new `PrestaShopBundle\Controller\Admin\PrestaShopAdminController`, which doesn't implement the mentioned hack to force a proper implementation of the controllers.
```

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -78,7 +78,91 @@ Some dependencies are still present but were upgraded which comes with their own
+This gives developer some time to refacto their controller, they should now rely on the new `PrestaShopBundle\Controller\Admin\PrestaShopAdminController` which doesn't implement this hack in order to force a proper implementation of the controllers.
+You can read more details about <a href="https://symfony.com/doc/6.4/controller/service.html" target="_blank">Controllers as services</a>, <a href="https://symfony.com/doc/6.4/service_container/service_subscribers_locators.html" target="_blank">Service Subscribers & Locators</a> or <a href="https://symfony.com/doc/6.4/service_container.html">Service container</a> in general.
+
+As an example you have three main possibilities to injected a service into your controller:
```
```suggestion
As an example, you have three main possibilities to inject a service into your controller:
```

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -78,7 +78,91 @@ Some dependencies are still present but were upgraded which comes with their own
+
+use PrestaShopBundle\Controller\Admin\PrestaShopAdminController;
+
+class MyController extends PrestaShopAdminController {
```
```suggestion
class MyController extends PrestaShopAdminController
{
```

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -78,7 +78,91 @@ Some dependencies are still present but were upgraded which comes with their own
+    }
+    
+    /**
+     * You can inject a service only on one method, useful if the service is used only in one action, it is more optimized than injecting it all the time and never use it.
```
```suggestion
     * You can inject a service only on one method, useful if the service is used only in one action. It is more optimized than injecting it all the time and never using it.
```

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -78,7 +78,91 @@ Some dependencies are still present but were upgraded which comes with their own
+    }
+    
+    /**
+     * You can define a list of registered services, they will be accessible via the $this->>container->get method, this
```
```suggestion
     * You can define a list of registered services, they will be accessible via the $this->container->get method, this
```

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -78,7 +78,91 @@ Some dependencies are still present but were upgraded which comes with their own
+    
+    /**
+     * You can define a list of registered services, they will be accessible via the $this->>container->get method, this
+     * way of injecting service is interesting if implemented on a base controller class for generic services that
```
I didn't know about it, it's indeed interesting 👀

jolelievre replied:
Yep it's quite elegant, it's used by the framework without us realizing it

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -78,7 +78,91 @@ Some dependencies are still present but were upgraded which comes with their own
+     */
+    public static function getSubscribedServices(): array
+    {
+        return return parent::getSubscribedServices() + [
```
```suggestion
        return parent::getSubscribedServices() + [
```

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -78,7 +78,91 @@ Some dependencies are still present but were upgraded which comes with their own
+    autowire: true
+    # Using autoconfigure with controllers that extend AbstractController will handle method parameter injection and subscribed services
+    autoconfigure: true
+    # If your controller class doesn't extend AbstractController you should add this tag manually si you benefit from method parameters injection
```
```suggestion
    # If your controller class doesn't extend AbstractController, you should add this tag manually, so you benefit from method parameters injection
```

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -78,7 +78,91 @@ Some dependencies are still present but were upgraded which comes with their own
+**PrestaShopAdminController::getSubscribedServices**
+
+The new base controller class uses the `getSubscribedServices` method to give easy access to some commonly used PrestaShop services, like the `ConfigurationInterface`, `HookDispatcherInterface`, `TranslatorInterface`, ...
+You can see the <a href="https://github.com/PrestaShop/PrestaShop/blob/develop/src/PrestaShopBundle/Controller/Admin/PrestaShopAdminController.php#L59" target="_blank">full list of subscribed services</a> in the class, this list will probably grow piece by piece but only for common generic services.
```
```suggestion
You can see the <a href="https://github.com/PrestaShop/PrestaShop/blob/9.0.0-alpha.1/src/PrestaShopBundle/Controller/Admin/PrestaShopAdminController.php#L59" target="_blank">full list of subscribed services</a> in the class, this list will probably grow piece by piece but only for common generic services.
```

we'll change to v9 final once ready

jolelievre replied:
Yeah that's what I though as well, that's why I used the branch for now But okay for that tag

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -78,7 +78,91 @@ Some dependencies are still present but were upgraded which comes with their own
+The new base controller class uses the `getSubscribedServices` method to give easy access to some commonly used PrestaShop services, like the `ConfigurationInterface`, `HookDispatcherInterface`, `TranslatorInterface`, ...
+You can see the <a href="https://github.com/PrestaShop/PrestaShop/blob/develop/src/PrestaShopBundle/Controller/Admin/PrestaShopAdminController.php#L59" target="_blank">full list of subscribed services</a> in the class, this list will probably grow piece by piece but only for common generic services.
+<br/><br/>
+Thanks to this you can use `$this->container->get(ConfigurationInterface::class)->get('my_config')` to access the configuration service, or alternatively use the helper method `$this->getConfiguration()->get('my_config')`.
```
```suggestion
Thanks to this, you can use `$this->container->get(ConfigurationInterface::class)->get('my_config')` to access the configuration service, or alternatively use the helper method `$this->getConfiguration()->get('my_config')`.
```

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -178,6 +262,29 @@ If you need to integrate your code with the authentication process you should no
 
+### Symfony context refacto
+
+Lots of core code is based on the legacy `Context` class that stores some data as a singleton, but this legacy implementation has many drawbacks so a refactoring was started in PrestaShop 9.0 to replace this usage.
```
```suggestion
A significant amount of code in the core relies on the legacy 'Context' class, which stores certain data as a singleton. However, this legacy implementation has many drawbacks, so refactoring was initiated in PrestaShop 9.0 to replace this usage.
```

jolelievre replied:
Ok except for `Context` that should be in backquotes IMHO, as it's a reference to some code

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -178,6 +262,29 @@ If you need to integrate your code with the authentication process you should no
+- `PrestaShop\PrestaShop\Core\Context\LegacyControllerContext`
+- `PrestaShop\PrestaShop\Core\Context\ShopContext`
+
+We still have some work to do to replace all the usages of the legacy context with the modern ones, but new code will stop using the legacy ones. We recommend developers to rely on the new ones from now on.
```
```suggestion
We still have some work to do to replace all the usages of the legacy context with the modern ones, but the new code will stop using the legacy ones. We recommend that developers rely on the new ones from now on.
```

kpodemski reviewed (commented):
Some suggestions.

I think you touched on some fundamental subjects that should go into the general documentation.

jolelievre commented:
Thanks for the review @kpodemski 
I addressed all your feedback I agree that this page defines many important changes that should probably be mentioned in other places As I proposed when I started documenting these changes, the idea is to do it all here for now, just because it was more convenient for me

But once everything is documented (only this PR and the one about Symfony layout left), we should indeed think about how to improve it:
- some parts deserve a dedicated page
- some parts should be copied in other relevant places probably
